crates/eip4844/src/verifier.rs

```diff
@@ -135,3 +139,117 @@ impl Context {
         Ok(())
     }
 }
+
+pub(crate) fn blob_scalar_to_polynomial(domain: &Domain, blob_scalar: &[Scalar]) -> PolyCoeff {
+    let mut polynomial = blob_scalar.to_vec();
+    bitreverse_slice(&mut polynomial);
+    domain.ifft_scalars(polynomial)
+}
+
+/// Compute Fiat-Shamir challenge of a blob KZG proof.
+///
+/// The matching function in the specs is: https://github.com/ethereum/consensus-specs/blob/017a8495f7671f5fff2075a9bfc9238c1a0982f8/specs/deneb/polynomial-commitments.md#compute_challenge
+pub(crate) fn compute_fiat_shamir_challenge(blob: BlobRef, commitment: KZGCommitment) -> Scalar {
+    // DomSepProtocol is a Domain Separator to identify the protocol.
+    //
+    // It matches [FIAT_SHAMIR_PROTOCOL_DOMAIN] in the spec.
+    //
+    // [FIAT_SHAMIR_PROTOCOL_DOMAIN]: https://github.com/ethereum/consensus-specs/blob/017a8495f7671f5fff2075a9bfc9238c1a0982f8/specs/deneb/polynomial-commitments.md#blob
+    const DOMAIN_SEP: &str = "FSBLOBVERIFY_V1_";
+
+    let bytes_per_commitment = G1Point::compressed_size();
+    let bytes_per_blob = blob.len();
+
+    let bytes_per_field_element = Scalar::NUM_BITS.div_ceil(8) as usize;
+    let field_elements_per_blob = blob.len() / bytes_per_field_element;
+
+    let hash_input_size = DOMAIN_SEP.len()
+            + 2 * size_of::<u64>() // polynomial bound
```

Comment:
Can we say what this 2 is for?

Reply:
It should've be `size_of::<u128>` I think, in the spec they serialize the polynomial bound to 16 bytes, will update in another PR.